### PR TITLE
feat: Add a cross-process lock mechanism for OIDC token refresh

### PR DESCRIPTION
**Pull request by bnjbvr**
This adds a cross-process lock for refresh to work correctly.

We want to coordinate token refresh across multiple processes. For that, we're using a cross-process lock, and a value in the database identifying the latest session tokens that are valid (a hash of the actual tokens, for security reasons).

Whenever we run into an HTTP error indicating that the tokens have been invalidated, we try to refresh the access tokens; that's already existing prior to this PR. The novelty introduced is that we take a cross-process lock before doing so, now. Taking this lock will also load a session hash from the database, and we'll compare it against the latest "known" session hash (that the current process saved into its memory).

If there's no mismatch (i.e. the database and the currently known are the same), then we're all good and can keep going with the refresh, synchronize the hashes everywhere (in-memory and database), make sure the client is notified about it (through a new user-provided callback `SaveSessionCallback`; on iOS this will save it into the device's keychain).

Otherwise, that means another process has done a refresh under our feet. In that case, we ask an authoritative source for trusted session tokens. On iOS, they're reloaded from the device keychain; that happens through a new user-provided callback `ReloadSessionCallback`. Then, we make sure that the DB and the in-memory value recall this latest value.

An embedder who would like to make use of the cross-process locking mechanism should call `client.oidc().set_session_callbacks` and `client.oidc().enable_cross_process_refresh`. If only interested with the pings for new sessions, the client may only call `client.oidc().set_session_callbacks`.

Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/2418.
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/2476

## Future improvements

- More testing of the whole flow. Not sure if mocking will be quite fit for OIDC, as this may require setting up an HTTPS server for the authentication code exchange and other OIDC-specific flows.
- Get rid of `SessionChange`, which duplicates in some way how a client can be notified about session changes.

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 1]
```diff
@@ -799,6 +723,42 @@ impl Client {
             });
         }
     }
+
+    fn retrieve_session(&self) -> Result<SessionTokens, String> {
+        let Some(delegate) = self.session_delegate.as_ref() else {
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client.rs:728]:
```suggestion
        let Some(delegate) = &self.session_delegate else {
```

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 2]
```diff
@@ -799,6 +723,42 @@ impl Client {
             });
         }
     }
+
+    fn retrieve_session(&self) -> Result<SessionTokens, String> {
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client.rs:727]: `String` seems strictly worse than `anyhow::Error`, any reason to use it here?

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 3]
```diff
@@ -799,6 +723,42 @@ impl Client {
...
+        let auth_api = self.inner.auth_api().context("Missing authentication API")?;
+
+        let homeserver_url = self.inner.homeserver().await.into();
+        let sliding_sync_proxy = self.discovered_sliding_sync_proxy().map(|url| url.to_string());
+
+        Session::new(auth_api, homeserver_url, sliding_sync_proxy)
+    }
+
+    async fn save_session(&self) -> Result<(), String> {
+        let Some(delegate) = self.session_delegate.as_ref() else {
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client.rs:755]:
```suggestion
        let Some(delegate) = &self.session_delegate else {
```

File: bindings/matrix-sdk-ffi/src/client_builder.rs [hunk 4]
```diff
@@ -43,6 +45,23 @@ impl ClientBuilder {
         Arc::new(Self::default())
     }
 
+    pub fn enable_cross_process_refresh_lock(
+        self: Arc<Self>,
+        process_id: String,
+        session_delegates: Box<dyn ClientSessionDelegate>,
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client_builder.rs:51]:
```suggestion
        session_delegate: Box<dyn ClientSessionDelegate>,
```

File: bindings/matrix-sdk-ffi/src/client_builder.rs [hunk 5]
```diff
@@ -34,6 +34,8 @@ pub struct ClientBuilder {
     disable_ssl_verification: bool,
     disable_automatic_token_refresh: bool,
     inner: MatrixClientBuilder,
+    cross_process_refresh_lock_id: Option<String>,
+    session_delegates: Option<Arc<dyn ClientSessionDelegate>>,
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client_builder.rs:38]:
```suggestion
    session_delegate: Option<Arc<dyn ClientSessionDelegate>>,
```

File: bindings/matrix-sdk-ffi/src/client_builder.rs [hunk 6]
```diff
@@ -116,6 +135,26 @@ impl ClientBuilder {
 }
 
 impl ClientBuilder {
+    pub(crate) fn enable_cross_process_refresh_lock_inner(
+        self: Arc<Self>,
+        process_id: String,
+        session_delegates: Arc<dyn ClientSessionDelegate>,
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client_builder.rs:141]:
```suggestion
        session_delegate: Arc<dyn ClientSessionDelegate>,
```

File: bindings/matrix-sdk-ffi/src/client_builder.rs [hunk 7]
```diff
@@ -116,6 +135,26 @@ impl ClientBuilder {
...
+    ) -> Arc<Self> {
+        let mut builder = unwrap_or_clone_arc(self);
+        builder.cross_process_refresh_lock_id = Some(process_id);
+        builder.session_delegates = Some(session_delegates);
+        Arc::new(builder)
+    }
+
+    pub(crate) fn set_session_delegates_inner(
+        self: Arc<Self>,
+        session_delegates: Arc<dyn ClientSessionDelegate>,
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client_builder.rs:151]:
```suggestion
        session_delegate: Arc<dyn ClientSessionDelegate>,
```

**Review by jplatte — COMMENTED** [review]
FFI review

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 2 above)
Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/client.rs]: Good point, I've enabled anyhow conditionally on OIDC, in the sdk crate.

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    #[derive(Hash)]
+    struct HashableSessionTokens {
+        access_token: String,
+        refresh_token: Option<String>,
+    }
+
+    SessionHash(hash(&HashableSessionTokens {
+        access_token: tokens.access_token.clone(),
+        refresh_token: tokens.refresh_token.clone(),
+    }))
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:37]:
```suggestion
    struct HashableSessionTokens<'a> {
        access_token: &'a str,
        refresh_token: Option<&'a str>,
    }

    SessionHash(hash(&HashableSessionTokens {
        access_token: &tokens.access_token,
        refresh_token: tokens.refresh_token.as_deref(),
    }))
```

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        Self { store, store_lock: lock, known_session_hash: Arc::new(Mutex::new(None)) }
+    }
+
+    /// Returns the value in the database representing the lock holder.
+    pub fn lock_holder(&self) -> &str {
+        self.store_lock.lock_holder()
+    }
+
+    /// Will wait for 60 seconds to get a cross-process store lock, then either
+    /// timeout (as an error) or return a lock guard.
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:59]:
```suggestion
    /// Wait for up to 60 seconds to get a cross-process store lock, then either
    /// timeout (as an error) or return a lock guard.
```

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+            .remove_custom_value(OIDC_SESSION_HASH_KEY)
+            .await
+            .map_err(CrossProcessRefreshLockError::StoreError)?;
+        *self.known_session_hash.lock().await = None;
+        Ok(())
+    }
+}
+
+pub(super) struct CrossProcessRefreshLockGuard {
+    /// The hash for the latest session, either we one we knew, or the latest
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:129]:
```suggestion
    /// The hash for the latest session, either the one we knew, or the latest
```

?

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    /// If so, this indicates that another process may have refreshed the token
+    /// in the background.
+    ///
+    /// We don't consider it a mismatch if there was no previous value in the
+    /// database. We do consider it a mismatch if there was no in-memory
+    /// value known, but one was known in the database.
+    pub hash_mismatch: bool,
+
+    /// Session hash previously stored in the DB. Used for debugging and testing
+    /// purposes.
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:151]:
```suggestion
    /// Session hash previously stored in the DB.
    ///
    /// Used for debugging and testing purposes.
```

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+    /// Updates the `SessionTokens` hash in both memory and database.
+    ///
+    /// Must be called after a successful refresh.
+    pub async fn save_in_memory_and_db(
+        &mut self,
+        tokens: &SessionTokens,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        let hash = compute_session_hash(tokens);
+        // Note: fallible operation at the end, so the in-memory value at least is up to
+        // date.
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:179]: Is a partial update really what we want if a full update fails?

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    /// Handle a mismatch by making sure values in the database and memory match
+    /// tokens we trust.
+    pub async fn handle_mismatch(
+        &mut self,
+        trusted_tokens: &SessionTokens,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        let new_hash = compute_session_hash(trusted_tokens);
+        trace!("Trusted OIDC tokens have hash {:?}; db had {:?}", new_hash, self.db_hash);
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:192]:
```suggestion
        trace!("Trusted OIDC tokens have hash {new_hash:?}; db had {:?}", self.db_hash);
```

File: crates/matrix-sdk/src/oidc/cross_process.rs [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+        if let Some(db_hash) = &self.db_hash {
+            if new_hash != *db_hash {
+                // That should never happen, unless we got into an impossible situation!
+                // In this case, we assume the value returned by the callback is always
+                // correct, so override that in the database too.
+                tracing::error!("error: DB and trusted disagree. Overriding in DB.");
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/cross_process.rs:199]: Why not import the `error` macro? Also, please don't prefix an error-level tracing event with `error: ` ^^

File: crates/matrix-sdk/src/oidc/mod.rs [hunk 15]
```diff
@@ -202,19 +204,55 @@ use rand::{rngs::StdRng, Rng, SeedableRng};
...
+};
 pub use self::{
     auth_code_builder::{OidcAuthCodeUrlBuilder, OidcAuthorizationData},
     end_session_builder::{OidcEndSessionData, OidcEndSessionUrlBuilder},
 };
 use crate::{authentication::AuthData, client::SessionChange, Client, RefreshTokenError, Result};
 
+type SaveSessionCallback =
+    dyn Send + Sync + Fn() -> Pin<Box<dyn Send + Sync + Future<Output = anyhow::Result<()>>>>;
+type ReloadSessionCallback = dyn Send + Sync + Fn() -> anyhow::Result<SessionTokens>;
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/mod.rs:227]: Having `Send` & `Sync` before the "main" trait of a `dyn` type seems unusual.

File: crates/matrix-sdk/src/oidc/mod.rs [hunk 16]
```diff
@@ -244,6 +282,84 @@ impl Oidc {
...
+            return Ok(());
+        };
+
+        // If the lock has already been created, don't recreate it from scratch.
+        if let Some(prev_lock) = self.ctx().cross_process_token_refresh_manager.get() {
+            let prev_holder = prev_lock.lock_holder();
+            if prev_holder == lock_value {
+                return Ok(());
+            }
+            warn!("recreating cross-process store refresh token lock with a different holder value: prev was {prev_holder}, new is {lock_value}");
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/mod.rs:342]: Overly long line, maybe use event fields?

```suggestion
            warn!(
                prev_holder, new_holder = lock_value,
                "recreating cross-process store refresh token lock with a different holder value"
            );
```

File: crates/matrix-sdk/src/oidc/mod.rs [hunk 17]
```diff
@@ -858,54 +1063,94 @@ impl Oidc {
...
+                    };
+
+                    this.set_session_tokens(tokens.clone());
+
+                    // Call the save_session_callback if set, while the optional lock is being held.
+                    if let Some(save_session_callback) = this.ctx().save_session_callback.get() {
+                        // Satisfies the save_session_callback invariant: set_session_tokens has
+                        // been called just above.
+                        if let Err(err) = save_session_callback().await {
+                            tracing::error!("error when saving session after refresh: {err}");
```
Comment (jplatte) [inline crates/matrix-sdk/src/oidc/mod.rs:1129]: Same as before, why not import?

File: crates/matrix-sdk/src/oidc/cross_process.rs (hunk 8 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/cross_process.rs:37]: Noice

File: crates/matrix-sdk/src/oidc/cross_process.rs (hunk 12 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/cross_process.rs:179]: I've removed that comment, as I suspect this mattered at some point during the development of the PR (i.e. just logging out an `Err` result if it failed, in a detached task), but now it's not the case anymore.

File: crates/matrix-sdk/src/oidc/cross_process.rs (hunk 13 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/cross_process.rs:192]: I'm not a huge fan of mixing inline field names with non-inlined, but I don't mind too much either.

File: crates/matrix-sdk/src/oidc/mod.rs (hunk 15 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/mod.rs:227]: Oh it's me being noobish-cautious and not knowing if the bound would apply to the result type of the function bound. But you're right, will fix.

**Review by jplatte — APPROVED** [review]
🚢

**Comment by bnjbvr** [comment]
Thanks for the review! I've fixed the cyclic reference issue you've identified during our chat (makes the FFI code a bit uglier, but oh well), and will keep the other improvements to follow-up PRs. Let's do an ultimate testing session and merge it.

**Pull request merged**